src/zero_copy/combinator.rs

```diff
@@ -1727,37 +1727,47 @@ where
     go_extra!(O);
 }
 
-// TODO: Finish implementing this once full error recovery is implemented
-/*#[derive(Copy, Clone)]
-pub struct Validate<A, F> {
+/// See [`Parser::validate`]
+pub struct Validate<A, OA, F> {
     pub(crate) parser: A,
     pub(crate) validator: F,
+    pub(crate) phantom: PhantomData<OA>,
+}
+
+impl<A: Copy, OA, F: Copy> Copy for Validate<A, OA, F> {}
+impl<A: Clone, OA, F: Clone> Clone for Validate<A, OA, F> {
+    fn clone(&self) -> Self {
+        Validate {
+            parser: self.parser.clone(),
+            validator: self.validator.clone(),
+            phantom: PhantomData,
+        }
+    }
 }
 
-impl<'a, I, O, E, S, A, F> Parser<'a, I, O, E, S> for Validate<A, F>
+impl<'a, I, OA, U, E, S, A, F> Parser<'a, I, U, E, S> for Validate<A, OA, F>
 where
     I: Input + ?Sized,
     E: Error<I>,
     S: 'a,
-    A: Parser<'a, I, O, E, S>,
-    F: Fn(E, I::Span, &mut dyn FnMut(E)) -> E,
+    A: Parser<'a, I, OA, E, S>,
+    F: Fn(OA, I::Span, &mut dyn FnMut(E)) -> U,
```

Comment:
I wonder whether we should get rid of the trait object in favour of a dedicated struct to avoid the dynamic dispatch... `Emitter<E>` and `emitter.emit(my_error)` :thinking:

Reply:
That doesn't sound horrible - and allows for future extensibility, if there's ever call for, say, custom error prioritization